Make file write operations more robust

matthijskooijman opened the pull request:
This changes file writes to go through a temporary file, which prevents data corruption when the write itself fails (e.g. due to disk full).

RoukePouw commented on system/core/database.php:
```diff
@@ -599,10 +599,40 @@ function __construct($filename) {
+			// older error when something goes wrong without
+			// generating a PHP error (e.g. like read/write
+			// before PHP 7.4).
+			error_clear_last();
```
Deze code en comment komen verderop nog een keer voor. Refactor kansje

matthijskooijman replied:
Mjah, je kunt deze code verplaatsen naar een aparte functie en die dan aanroepen, maar dan vervang je 1 functiecall door 1 functiecall. Je kunt eventueel de comment dan weglaten, maar eigenlijk hoort dit ocmment niet *in* die functie, maar hier, bij de aanroep, omdat de context is waar de comment zinnig in is.

Of had je een andere refactor in gedachten?

RoukePouw replied:
Nee, nu je het zegt. Hierom refactor is wellicht wat omslachtig, ik werd alleen getriggerd door twee keer dezelfde tekst :)

RoukePouw commented on system/core/database.php:
```diff
@@ -599,10 +599,40 @@ function __construct($filename) {
 		function lock() {
 			if ($this->fd)
```
Terzijde opmerking: de naamte `fd` maakte het voor mij niet direct duidelijk waar dit om ging, dat kan wellicht verduidelijkt worden :)

matthijskooijman replied:
fd slaat op file descriptor, wat iig in C-code wel een veel gebruikte naam is. Ik zie dat de fd variable niet gedeclareerd is, dus dat kan ik wel even doen met een commentje erbij.

matthijskooijman replied:
Oh wacht, ik zat bij `HyphaImage` te kijken, in `HyphaFile` is `$fd` al wel gedeclareerd is met een commentje (wel met foutje, null vs false). Blijkbaar is dat nog niet voldoende om het duidelijk te maken :-p

Kijkende naar [de docs](https://www.php.net/manual/en/function.fopen.php) is het eigenlijk ook geen file descriptor (dat zijn ints, e.g. 0 voor stdin, 1 voor stdout, 2 voor stderr en 3+ voor opened files) die teruggeven wordt, maar een of ander PHP-specifiek resource object, dus is een andere naam misschien wel een goed idee.

Ik denk aan `$file_handle`, zou dat voor jouw duidelijker zijn geweest?

RoukePouw replied:
`$file_handle` = top :)

matthijskooijman replied:
Done.

RoukePouw commented on system/core/database.php:
```diff
@@ -599,10 +599,40 @@ function __construct($filename) {
+			// before PHP 7.4).
+			error_clear_last();
+
+			for ($attempt = 0; $attempt < 10; ++$attempt) {
```
Magic number, overweeg deze als `MAX_NUMBER_ATTEMPTS` oid te implementeren

matthijskooijman replied:
Mjah, het is een waarde die heel lokaal is, dus er een class constant van maken is wellicht wat overdone, maar het is wel een magic number... Ik heb er nu maar een lokale "constante" van gemaakt, dan heeft het iig een naam. Hoe lijkt dat?

RoukePouw replied:
Perfect

RoukePouw commented on system/core/database.php:
```diff
@@ -599,10 +599,40 @@ function __construct($filename) {
+				// file. Note: This makes all locking
+				// fail when the file is hardlinked
+				// elsewhere.
+				if (fstat($fd)['nlink'] != 0) {
```
Wat is het standpunt wat betreft type juggling?

matthijskooijman replied:
Wat bedoel je hier precies? Dat we eigenlijk `!==` zouden moeten gebruiken?

RoukePouw replied:
ik wilde het aandragen als suggestie en was benieuwd naar jouw standpunt hierover. Voor mij "springt" `!=` er altijd uit in js en php. In c(++) onderdruk ik de neiging om mezelf er van te laten opkijken ;-)

matthijskooijman replied:
Ik programmeer meer in C/C++, waar dit niet zo'n issue is (en !== uberhaupt niet bestaat), dus ik ben er niet zo op bedacht :-) Ik geloof dat Laurens waar mogelijk ook stricte checks gebruikt, en hier lijkt het me ook eigenlijk wel een goed idee.

OTOH, stel dat er een keer `null` oid uit zou komen (kan nu volgens mij niet, maar wie weet is er nog een keer een platform waarop dat wel gebeurt), dan zou bij `!== 0` de if triggeren, terwijl er eigenlijk geen (of niet zeker) een geldig lock is. Dus eigenlijk juist beter als dan de *if* niet triggert (en als dat dan 10x hetzelfde is, krijg je een exception, wat wsl beter is dan geen lock terwijl je denkt dat je er wel 1 hebt). Ik heb er nu `> 0` van gemaakt, dat heeft hetzelfde effect, maar is misschien nog iets duidelijker (het is eigenlijk niet 0 die speciaal is, maar > 0).

RoukePouw replied:
Kan ik me helemaal in vinden

RoukePouw commented on system/core/database.php:
```diff
@@ -659,24 +681,86 @@ function lockAndRead() {
+			// problems, which would break atomicity if it
+			// is on another filesystem.
+			//
+			$tmpfilename = $this->filename . '.tmp_for_writing';
```
is het een idee om hier een `uniqid()` achteraan te plakken?

matthijskooijman replied:
Heb ik over nagedacht, maar in principe zou het lock op de file zelf moeten zorgen dat er nooit twee tegelijk naar dezelfde file schrijven. Mocht er toch ooit een keer een tmpfile blijven rondslingeren heeft *geen* uniqid toevoegen als voordeel dat die tmpfile vanzelf opgeruimd wordt bij een volgende write. Klinkt dat zinnig?

RoukePouw replied:
Klinkt zinnig :-)

RoukePouw commented on system/core/database.php:
```diff
@@ -659,24 +681,86 @@ function lockAndRead() {
+			//    writing to the file directly).
+			//
+			// This uses a fixed temp filename instead of
+			// using tempnam, since the latter might default
```
typo `tempnam, `

matthijskooijman replied:
Nope, die functie heet echt zo: https://www.php.net/manual/en/function.tempnam.php

matthijskooijman commented:
I've rebased on top of master, squashed the fixup commit and also did one more rename from `$file_handle` to `$fileHandle` to actually conform to the naming convention used. With that, I think this is ready to merge, so I'll merge this.
